### DISCUSSION TITLE
Building a custom Airflow base image with other linux distributions

### DISCUSSION THREAD

**Discussion by lnfrarad** [apache/airflow#36295] [Q&A] [answered]
Hi folks, 

In my company we are using Airflow, and recently the Security dept has started reviewing our vulnerability scans. Unfortunately Debian on the current base image is triggering a lot of issues with our scanner. No doubt these could mostly be false positives but no one on our dev team currently has the expertise to write them off as such. 

The recent Airflow 2.8 release on Debian bookworm helps to lower the vulnerability count. Thanks so much to the contributors for the work there. 

To work around the issue further, we would like to try building a custom image with different linux distribution eg: ubuntu or a newer Debian version. Ive seen the Airflow documentation with suggestion on how to [customize the image](https://airflow.apache.org/docs/docker-stack/build.html#extending-vs-customizing-the-image). 

Also @potiuk mentioned in a separate airflow issue [#31941: Switch Airlfow base images to use debian Bookworm ](https://github.com/apache/airflow/issues/31941) that new official image builds need to go through airflow Docker CI, and be verified to work with 80 providers before it can be released. 

Can someone pls direct me to the test cases which are used for verification, so that I can verify our custom image as well. And to make the right choice to stick with the official image if a custom image has too many incompatibilities. Thank you.

**Comment by raphaelauv** [comment]
you create your own "at your own risk"

```dockerfile
FROM python:3.11-slim

ENV AIRFLOW_HOME=/usr/local/airflow

ENV PIP_NO_CACHE_DIR=TRUE
ENV PIP_DISABLE_PIP_VERSION_CHECK=TRUE

RUN  useradd -ms /bin/bash -d ${AIRFLOW_HOME} airflow \
    && python3 -m pip install --upgrade pip setuptools


RUN python3 -m pip install apache-airflow==2.8.0 --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.8.0/constraints-3.11.txt"

COPY entrypoint.sh /

RUN chown -R airflow: ${AIRFLOW_HOME}

WORKDIR ${AIRFLOW_HOME}
ENTRYPOINT ["/entrypoint.sh"]
```

**Reply by lnfrarad** [reply]
Hey @raphaelauv thanks for the sample dockerfile. :thumbsup:

**Comment by potiuk** [comment, marked as answer]
Yep. Good one @raphaelauv  - there is nothing super special about Airflow image except that it is complex and versatile and serves many purposes (including being used in CI, handling Redshift best practics, being heavily optimized and about a dozen of other things) while you can certainly have a simple "do-it-yourself" image, where you will likely have to solve some problems on your own (say for example MySQL changing their signing keys which happend last week)  - and if you need other distros  as a base, you can definitely follow that route.

Regarding the tests - we do have a few tests that we are running when we verify the image. They are targetted for our image, so it might or might not work if you build your own do-it-yourself image. We are using them in CI after the image is built to verify that it has what we want it to have. Those tests are rudimentary - they basically check if some simple commands (like "airflow version") work whether some packages are installed and few other things (like the Redshif compliance with running the image as root and custom user). 

You will have to likely move them out of the Airflow source tree and adapt, and solve some of the incompatibllities, but you can actually take them as a base and likely adapt to your needs. 

https://github.com/apache/airflow/blob/main/docker_tests/test_prod_image.py

They were not foreseen to test "other" images so you will have to find your own way and adapt them if you want to use similar approach.

**Reply by lnfrarad** [reply]
Hi @potiuk thanks for the pointers, i'll check it out this weekend.  :ok_hand:
